### PR TITLE
feat: add hidden properties

### PR DESCRIPTION
**Pull request by mdbetancourt**
Implements: #6385 
Related to: #1914

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] Affected artifact templates in `packages/cli` were updated
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/repository/src/__tests__/unit/model/model.unit.ts [hunk 1]
```diff
@@ -46,8 +46,7 @@ describe('model', () => {
   const flexibleDef = new ModelDefinition('Flexible');
   flexibleDef
     .addProperty('id', {type: 'string', id: true})
-    .addProperty('createdAt', Date)
-    .addSetting('hiddenProperties', ['createdAt'])
+    .addProperty('createdAt', {type: Date, hidden: true})
```
Comment (bajtos) [inline packages/repository/src/__tests__/unit/model/model.unit.ts:49]: Please don't modify existing tests, we want to keep this test to ensure `hiddenProperties` setting is still working as expected. Add a new test and/or define a new test model instead.

In my experience, it's best to avoid sharing test model that's very specific and define such model inside the test that is using it, see `it('includes navigational properties in JSON` below for an example to follow.

File: packages/repository/src/model.ts [hunk 2]
```diff
@@ -40,6 +40,7 @@ export type PropertyType =
 export interface PropertyDefinition {
   type: PropertyType; // For example, 'string', String, or {}
   id?: boolean | number;
+  hidden?: boolean;
```
Comment (bajtos) [inline packages/repository/src/model.ts:43]: Please add a TSDoc comment to explain the purpose of this property and mention how it interacts with model-level setting `hiddenProperties`.

```suggestion
  /**
   * (put the text here)
   */
  hidden?: boolean;
```

File: packages/repository/src/model.ts [hunk 3]
```diff
@@ -329,9 +330,9 @@ export abstract class Model {
 
     const json: AnyObject = {};
     const hiddenProperties: string[] = def.settings.hiddenProperties || [];
-    for (const p in def.properties) {
-      if (p in this && !hiddenProperties.includes(p)) {
-        copyPropertyAsJson(p);
+    for (const [pKey, pDef] of Object.entries(def.properties)) {
```
Comment (bajtos) [inline packages/repository/src/model.ts:333]: Perhaps `propName` and `propDef` would be better names? In `def.properties` map, entries are keyed using the property name.

```suggestion
    for (const [propName, propDef] of Object.entries(def.properties)) {
```

**Review by bajtos — CHANGES_REQUESTED** [review]
Thank you @mdbetancourt for the pull request! I have few comments on what to improve, see below.

I am little bit concerned that now we have to update every code dealing with hidden properties, are you sure you found all places? For example, `hiddenProperties` setting is passed to loopback-datasource-juggler, but properties with `hidden: true` are not. I am not able to tell if that could be a problem or not.

Have you considered a different approach, where properties with `hidden: true` are handled by model definition object and copied to `hiddenProperties` setting? That way we don't need to modify any code dealing with hidden properties, just the code building model definitions.

@raymondfeng @hacksparrow do you have any opinions on this?

**Comment by bajtos** [comment]
Also please update the documentation in [Model >> Hidden properties](https://github.com/strongloop/loopback-next/blob/628720c10ed638a884f547dd604bb88cbd297ad9/docs/site/Model.md#hidden-properties) to explain both ways how to hide a property.

**Comment by hacksparrow** [comment]
> Have you considered a different approach, where properties with `hidden: true` are handled by model definition object and copied to `hiddenProperties` setting? That way we don't need to modify any code dealing with hidden properties, just the code building model definitions.

Yes, that's what I'd prefer too, `{ hidden: true }` as a sugar for pushing the property to `hiddenProperties`.  As such `{ hidden: true }` should not be involved in any logic.

@mdbetancourt please consider this approach. It will be less complex and less disruptive.

File: packages/repository/src/__tests__/unit/model/model.unit.ts [hunk 4]
```diff
@@ -444,6 +444,31 @@ describe('model', () => {
     });
   });
 
+  it('hide a propety with PropertyDefinition', () => {
```
Comment (hacksparrow) [inline packages/repository/src/__tests__/unit/model/model.unit.ts:447]: `hide a propety with PropertyDefinition` -> `hides a property defined as hidden`

File: packages/repository/src/__tests__/unit/model/model.unit.ts [hunk 5]
```diff
@@ -444,6 +444,31 @@ describe('model', () => {
...
+    class UserHidden extends Entity {
```
Comment (hacksparrow) [inline packages/repository/src/__tests__/unit/model/model.unit.ts:448]: `UserHidden` -> `User`

**Comment by hacksparrow** [comment]
With the approach suggested by @bajtos and some minor text changes, this should be good to land.

File: packages/repository/src/__tests__/unit/model/model.unit.ts (hunk 5 above)
Comment (mdbetancourt) [reply packages/repository/src/__tests__/unit/model/model.unit.ts:448]: already exists a class called User to avoid overwrite the existed one i used that name

File: packages/repository/src/__tests__/unit/model/model.unit.ts (hunk 4 above)
Comment (mdbetancourt) [reply packages/repository/src/__tests__/unit/model/model.unit.ts:447]: > @mdbetancourt please consider this approach. It will be less complex and less disruptive.

yup i already used that approach
> hide a propety with PropertyDefinition -> hides a property defined as hidden

got it

File: docs/site/Model.md [hunk 6]
```diff
@@ -459,6 +460,21 @@ class MyUserModel extends Entity {
...
+
+```ts
+@model()
+class MyUserModel extends Entity {
+  @property({id: true})
+  id: number;
+   @property({type: 'string'})
+  email: string;
+   @property({type: 'string', hidden: true})
+  password: string;
```
Comment (bajtos) [inline docs/site/Model.md:473]: Indentation seems to be a bit off here.

```suggestion
  @property({type: 'string'})
  email: string;
  @property({type: 'string', hidden: true})
  password: string;
```

File: packages/repository/src/__tests__/unit/model/model.unit.ts (hunk 5 above)
Comment (bajtos) [reply packages/repository/src/__tests__/unit/model/model.unit.ts:448]: Can we use a class/model name like `Customer`?

Personally, I don't mind `UserHidden` too much, since it's scoped to this single test only.

File: packages/repository/src/model.ts (hunk 2 above)
Comment (bajtos) [inline packages/repository/src/model.ts:43]: I am afraid you haven't addressed my comment here, I am re-posting it again.

Please add a TSDoc comment to explain the purpose of this property and mention how it interacts with model-level setting `hiddenProperties`.

```suggestion
  /**
   * (put the text here)
   */
  hidden?: boolean;
```

File: packages/repository/src/model.ts [hunk 7]
```diff
@@ -329,9 +330,9 @@ export abstract class Model {
...
+      if (pKey in this && !hiddenProperties.includes(pKey) && !pDef.hidden) {
+        copyPropertyAsJson(pKey);
```
Comment (bajtos) [inline packages/repository/src/model.ts:335]: Since you are not changing the behavior here, please revert this block back to original version to avoid code churn and make the future code archaeology easier.

File: packages/repository/src/model.ts [hunk 8]
```diff
@@ -358,10 +359,11 @@ export abstract class Model {
     const def = (this.constructor as typeof Model).definition;
     const obj: AnyObject = {};
 
+    const props = def?.properties;
     if (options?.ignoreUnknownProperties === false) {
       const hiddenProperties: string[] = def?.settings.hiddenProperties || [];
       for (const p in this) {
-        if (!hiddenProperties.includes(p)) {
+        if (!hiddenProperties.includes(p) && !props?.[p]?.hidden) {
```
Comment (bajtos) [inline packages/repository/src/model.ts:366]: Can we revert this change please? I believe we agreed to handle `hidden` flag in the model definition only.

It would be best to revert all changes in `toJSON` method.

**Review by bajtos — COMMENTED** [review]
The new version looks much better 👍🏻 

I have few more comments to address, PTAL.

File: packages/repository/src/model.ts (hunk 2 above)
Comment (mdbetancourt) [reply packages/repository/src/model.ts:43]: i dont know what happen I remember doing this and removing the old code.

File: packages/repository/src/model.ts (hunk 8 above)
Comment (mdbetancourt) [reply packages/repository/src/model.ts]: yes of course

File: packages/repository/src/__tests__/unit/decorator/model-and-relation.decorator.unit.ts [hunk 9]
```diff
@@ -132,6 +132,24 @@ describe('model decorator', () => {
...
+      constructor(data: Partial<Client>) {
+        super(data);
+      }
+    }
+
+    expect(Client.definition.settings.hiddenProperties).to.containEql(
+      'password',
+    );
+  });
+
```
Comment (hacksparrow) [inline packages/repository/src/__tests__/unit/decorator/model-and-relation.decorator.unit.ts:152]: We should keep your first assertion too, which checked the `password` property is not present in the instance of `Client`, to make sure the intended functionality actually works.

**Comment by hacksparrow** [comment]
With the previous assertion also added and `UserHidden` renamed to `Customer`, this looks ready to land for me.

**Comment by dhmlau** [comment]
@hacksparrow, assigning this PR to you. Please help it to land. Thanks!

**Comment by hacksparrow** [comment]
@mdbetancourt the update looks good. However, we will have to add some more tests to address the drop in test coverage. Writing tests which cover any newly introduced `if`-`else` conditions will fix it.

Eg: Add a test case to confirm the property is not hidden if it does not have `hidden` set to `true`. Basically an inverse of your existing test case.

One may wonder, "What's the point of doing this?" This will ensure, the latest changes does not accidentally hide any property - now or in future.

**Review by bajtos — APPROVED** [review]
Code changes LGTM.

The drop in code coverage looks like a bug in coveralls infrastructure to me. If you look at the coverage report, it complains about missing coverage for an existing block which didn't have coverage before either.

I don't have a strong opinion on whether we need to add a new test to verify that properties with no `hidden` flag are not added to `hiddenProperties` array - I would expect this case is already covered by existing tests. Maybe not? I'll leave it up to @hacksparrow to decide.

@mdbetancourt can you please rebase your patch on top of the latest `master`, to see if it fixes the unexpected coverage drop? While doing that, please squash all commits into a single one and provide a descriptive error message. See our [commit message guidelines](https://loopback.io/doc/en/lb4/code-contrib-lb4.html#commit-message-guidelines) and [final rebase and squashing of commits](https://loopback.io/doc/en/lb4/submitting_a_pr.html#9-final-rebase-and-squashing-of-commits) if you are not familiar with the process.

**Comment by mdbetancourt** [comment]
> I don't have a strong opinion on whether we need to add a new test to verify that properties with no `hidden` flag are not added to `hiddenProperties` array - I would expect this case is already covered by existing tests. Maybe not? I'll leave it up to @hacksparrow to decide.

yes is already covered by a lot of test (almost all test doesn't use that flag

> The drop in code coverage looks like a bug in coveralls infrastructure to me. If you look at the coverage report, it complains about missing coverage for an existing block which didn't have coverage before either.
i added a test to cover this (i did a rebase and only left the feature and this fix)

File: packages/repository/src/__tests__/unit/decorator/model-and-relation.decorator.unit.ts [hunk 10]
```diff
@@ -132,6 +132,41 @@ describe('model decorator', () => {
...
+      name: 'name',
+    });
+  });
+  it('Throw error if desing type is not provided', () => {
+    const createModel = () => {
+      @model()
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      class Client extends Entity {
+        @property()
+        id: undefined;
```
Comment (hacksparrow) [inline packages/repository/src/__tests__/unit/decorator/model-and-relation.decorator.unit.ts:164]: Could there be a more practical way to write this test case?

Comment (mdbetancourt) [reply packages/repository/src/__tests__/unit/decorator/model-and-relation.decorator.unit.ts:164]: sorry i'm not sure what else i could use, i created #6522 related to this (only undefined, null and maybe emitDecoratorMetadata to false throws undefined)

Comment (hacksparrow) [reply packages/repository/src/__tests__/unit/decorator/model-and-relation.decorator.unit.ts:164]: Alright, we'll have to use a single commit message so please squash the two into one. Ready to land.

File: packages/repository/src/decorators/model.decorator.ts [hunk 11]
```diff
@@ -32,7 +32,7 @@ export const MODEL_WITH_PROPERTIES_KEY = MetadataAccessor.create<
   ClassDecorator
 >('loopback:model-and-properties');
 
-export type PropertyMap = MetadataMap<PropertyDefinition>;
+export type PropertyMap = MetadataMap<Partial<PropertyDefinition>>;
```
Comment (bajtos) [inline packages/repository/src/decorators/model.decorator.ts:35]: This change seems a bit unrelated to the test being added, I am also not sure if is a good idea to allow only partial property definitions in `PropertyMap`. Shouldn't we (the `@model` decorator?) guarantee that `PropertyMap` always contains valid (full) property definitions only?

File: packages/repository/src/__tests__/unit/decorator/model-and-relation.decorator.unit.ts [hunk 12]
```diff
@@ -132,6 +132,41 @@ describe('model decorator', () => {
...
+      password: 'password',
+    });
+    expect(Client.definition.settings.hiddenProperties).to.containEql(
+      'password',
+    );
+    expect(client.toJSON()).to.eql({
+      name: 'name',
+    });
+  });
+  it('Throw error if desing type is not provided', () => {
```
Comment (bajtos) [inline packages/repository/src/__tests__/unit/decorator/model-and-relation.decorator.unit.ts:158]: We write test names to read as English sentences. Also please add an empty line before the test case to serve as a visual delimiter.

```suggestion

  it('throws error if design type is not provided', () => {
```

File: packages/repository/src/decorators/model.decorator.ts (hunk 11 above)
Comment (mdbetancourt) [reply packages/repository/src/decorators/model.decorator.ts:35]: only "type" is required (but is inferred when is not provided)

https://github.com/strongloop/loopback-next/blob/11267618b58d437cd9a3c28f4c2f4f48cfe930c0/packages/repository/src/decorators/model.decorator.ts#L87-L89

here designType does not provide a properly type (string | Function | undefined) instead provide (string | Function) because `propDef.type` is not nullable, when in fact it is (that's why  we check and throw CANNOT_INFER_PROPERTY_TYPE error). i'm open to use another approach thought. like only change
https://github.com/strongloop/loopback-next/blob/11267618b58d437cd9a3c28f4c2f4f48cfe930c0/packages/repository/src/decorators/model.decorator.ts#L83

Comment (bajtos) [reply packages/repository/src/decorators/model.decorator.ts:35]: I though that `PropertyMap` is being used by `ModelDefinition` too. I did a [quick search](https://github.com/strongloop/loopback-next/search?q=PropertyMap) and see that this type is used only for decorator-related metadata, in which case this change is fine with me. (I was worrying about changing ModelDefinition contract, which is not the case.)

**Review by bajtos — APPROVED** [review]
LGTM 👍🏻 

@hacksparrow Would you like to take another look? Can we merge this PR?

**Comment by bajtos** [comment]
Landed, thank you @mdbetancourt for the contribution! ❤️
